### PR TITLE
Implement bid-silence and expiration when live-bidding

### PR DESCRIPTION
**Pull request by Amokrane**

### CODE REVIEW

**Comment by Amokrane** [comment]
@TurpIF @vguerci  note that there are two commits on this PR, each can be reviewed separately.

**Review by TurpIF — CHANGES_REQUESTED** [review]

File: publisher-sdk/src/main/java/com/criteo/publisher/BidManager.java [hunk 1]
```diff
@@ -204,14 +201,37 @@ private void fetchForLiveBidRequest(
       return;
     }
 
+    CdbResponseSlot cachedCdbResponseSlot;
+    synchronized (cacheLock) {
```
[the remark this answers is not in the document]
Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/BidManager.java:206]: Yep good point, done.

File: publisher-sdk/src/main/java/com/criteo/publisher/LiveCdbCallListener.java [hunk 2]
```diff
@@ -81,15 +81,20 @@ public void onCdbResponse(
         CdbResponseSlot cdbResponseSlot = cdbResponse.getSlots().get(0);
         if (!bidManager.isBidSilent(cdbResponseSlot)) {
           bidListener.onBidResponse(cdbResponseSlot);
+          bidLifecycleListener.onBidConsumed(cacheAdUnit, cdbResponseSlot);
         } else {
-          bidManager.setCacheAdUnits(cdbResponse.getSlots());
+          bidManager.setCacheAdUnits(
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/LiveCdbCallListener.java:86]: Nit: that's a lot of new lines :D

File: publisher-sdk/src/main/java/com/criteo/publisher/LiveCdbCallListener.java [hunk 3]
```diff
@@ -81,15 +81,20 @@ public void onCdbResponse(
...
+              cdbResponse.getSlots()
+          );
           bidListener.onNoBid();
         }
       } else {
         bidListener.onNoBid();
       }
     } else {
-      bidManager.setCacheAdUnits(cdbResponse.getSlots());
+      bidManager.setCacheAdUnits(
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/LiveCdbCallListener.java:95]: Nit: same

File: publisher-sdk/src/main/java/com/criteo/publisher/BidManager.java [hunk 4]
```diff
@@ -245,18 +246,23 @@ private void sendBidRequest(List<CacheAdUnit> prefetchCacheAdUnits) {
   }
 
   void setCacheAdUnits(@NonNull List<CdbResponseSlot> slots) {
+    List<CdbResponseSlot> cachedCdbResponseSlots = new ArrayList<>();
```
Comment (TurpIF) [inline publisher-sdk/src/main/java/com/criteo/publisher/BidManager.java:249]: What is this list used for ?

Comment (Amokrane) [reply publisher-sdk/src/main/java/com/criteo/publisher/BidManager.java:249]: Oops, left over from an initial direction that I gave up on later. REmoved.

**Review by TurpIF — APPROVED** [review]

**Pull request merged**